File: sdks/python/pasm_client.py
"""pasm-framework Python 客户端（零依赖，标准库 urllib）。

**Python 项目通常不需要这个文件** —— 同进程直接
``from pasm_framework import SimpleApplication`` 更快，还能用到
``Capability`` / ``DomainAdapter`` 等高级表面。
本客户端用于：跨进程/跨机调用、或对接**别人的** pasm-framework 服务。

用法::

    from pasm_client import PasmClient, PasmError

    c = PasmClient("http://127.0.0.1:8080", token="your-secret")
    print(c.chat("怎么退货？", session_id="user-1"))
    c.ingest([{"title": "退货政策", "content": "7 天内无理由退货。", "source": "faq"}])
    print(c.kb_stats())
    print(c.health())
"""
from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Dict, List, Optional


class PasmError(Exception):
    """统一的调用错误。

    ``status`` 为 HTTP 状态码（网络层失败时为 0）；
    ``hint`` 是服务端给出的修复建议（若提供）。
    """

    def __init__(self, message: str, status: int = 0, hint: str = "") -> None:
        super().__init__(message)
        self.status = status
        self.hint = hint

    def __str__(self) -> str:  # pragma: no cover
        base = super().__str__()
        return "%s（HTTP %s）%s" % (base, self.status,
                                   ("  提示：%s" % self.hint) if self.hint else "")


class PasmClient:
    """pasm-framework HTTP 网关的客户端。"""

    def __init__(self, base_url: str = "http://127.0.0.1:8080",
                 token: Optional[str] = None, timeout: float = 30.0) -> None:
        self.base_url = (base_url or "").rstrip("/")
        self.token = token or ""
        self.timeout = timeout

    # ---- 底层 ----
    def _request(self, method: str, path: str,
                 payload: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        url = self.base_url + path
        data = None
        headers = {"Accept": "application/json"}
        if payload is not None:
            data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
            headers["Content-Type"] = "application/json"
        if self.token:
            headers["Authorization"] = "Bearer " + self.token
        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                body = resp.read().decode("utf-8") or "{}"
            out = json.loads(body)
            return out if isinstance(out, dict) else {"data": out}
        except urllib.error.HTTPError as ex:
            raw = ex.read().decode("utf-8", "ignore") or "{}"
            try:
                obj = json.loads(raw)
            except Exception:
                obj = {"error": raw[:200]}
            raise PasmError(obj.get("error", "请求失败"), ex.code,
                            obj.get("hint", "")) from ex
        except urllib.error.URLError as ex:
            raise PasmError("无法连接 pasm-framework 服务：%s" % ex.reason, 0,
                            "确认服务已启动、地址与端口正确") from ex

    # ---- 业务 ----
    def chat(self, text: str, session_id: str = "default",
             user_id: Optional[str] = None,
             meta: Optional[Dict[str, Any]] = None) -> str:
        """发一条消息，返回回复文本。"""
        payload: Dict[str, Any] = {"text": text, "session_id": session_id}
        if user_id is not None:
            payload["user_id"] = user_id
        if meta is not None:
            payload["meta"] = meta
        return self._request("POST", "/api/chat", payload).get("reply", "")

    def chat_stream(self, text: str, session_id: str = "default",
                    user_id: Optional[str] = None,
                    meta: Optional[Dict[str, Any]] = None):
        """流式发送，逐条产出事件字典。

        事件：``{"type":"delta","text":…}`` / ``{"type":"replace","text":…}`` /
        ``{"type":"done",…}`` / ``{"type":"error",…}``。

        用法::

            acc = ""
            for ev in c.chat_stream("你好"):
                if ev["type"] == "delta":
                    acc += ev["text"]; print(ev["text"], end="", flush=True)
                elif ev["type"] == "replace":     # 护栏改写过内容 → 整条替换
                    acc = ev["text"]
        """
        payload: Dict[str, Any] = {"text": text, "session_id": session_id}
        if user_id is not None:
            payload["user_id"] = user_id
        if meta is not None:
            payload["meta"] = meta
        url = self.base_url + "/api/chat/stream"
        data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
        headers = {"Content-Type": "application/json",
                   "Accept": "text/event-stream"}
        if self.token:
            headers["Authorization"] = "Bearer " + self.token
        req = urllib.request.Request(url, data=data, headers=headers, method="POST")
        try:
            resp = urllib.request.urlopen(req, timeout=self.timeout)
        except urllib.error.HTTPError as ex:
            raw = ex.read().decode("utf-8", "ignore") or "{}"
            try:
                obj = json.loads(raw)
            except Exception:
                obj = {"error": raw[:200]}
            raise PasmError(obj.get("error", "请求失败"), ex.code,
                            obj.get("hint", "")) from ex
        except urllib.error.URLError as ex:
            raise PasmError("无法连接 pasm-framework 服务：%s" % ex.reason, 0,
                            "确认服务已启动、地址与端口正确") from ex
        # 逐行读 SSE（服务端以 EOF 结束流，不用 Content-Length）。
        with resp:
            for raw in resp:
                line = raw.decode("utf-8", "ignore").strip()
                if not line.startswith("data:"):
                    continue
                chunk = line[5:].strip()
                if not chunk:
                    continue
                try:
                    yield json.loads(chunk)
                except Exception:
                    continue

    def ingest(self, items: List[Dict[str, Any]]) -> int:
        """批量写入资料库，返回新增条数（需服务端启用 knowledge_base）。"""
        return int(self._request("POST", "/api/ingest", {"items": items})
                   .get("added", 0))

    def reset_session(self, session_id: str) -> bool:
        return bool(self._request("POST", "/api/sessions/reset",
                                  {"session_id": session_id}).get("ok"))

    def kb_stats(self) -> Dict[str, Any]:
        return self._request("GET", "/api/kb/stats")

    def plugins(self) -> Dict[str, Any]:
        return self._request("GET", "/api/plugins")

    def summary(self) -> Dict[str, Any]:
        return self._request("GET", "/api/summary")

    def health(self) -> Dict[str, Any]:
        """健康检查（免鉴权）。"""
        return self._request("GET", "/healthz")


if __name__ == "__main__":  # 手工冒烟
    import sys
    cli = PasmClient(sys.argv[1] if len(sys.argv) > 1 else "http://127.0.0.1:8080",
                     sys.argv[2] if len(sys.argv) > 2 else None)
    print("health:", cli.health().get("status"))
    print("reply :", cli.chat("怎么退货？"))

File: sdks/node/pasm-client.js
/**
 * pasm-framework Node.js 客户端（零依赖，用内置 fetch，需 Node 18+）。
 *
 * 用法：
 *   const { PasmClient, PasmError } = require("./pasm-client");
 *
 *   const c = new PasmClient("http://127.0.0.1:8080", "your-secret");
 *   console.log(await c.chat("怎么退货？", { sessionId: "user-1" }));
 *   await c.ingest([{ title: "退货政策", content: "7 天内无理由退货。", source: "faq" }]);
 *   console.log(await c.kbStats());
 *
 * 也可用作 Express/Koa 的 BFF 中间层，或 Electron/Tauri 桌面端的大脑客户端。
 */

class PasmError extends Error {
  constructor(message, status = 0, hint = "") {
    super(message);
    this.name = "PasmError";
    this.status = status;
    this.hint = hint;
  }
}

class PasmClient {
  /**
   * @param {string} baseUrl 例如 http://127.0.0.1:8080
   * @param {string|null} token 服务端配置的令牌；未配置则传 null
   * @param {number} timeout 超时毫秒
   */
  constructor(baseUrl = "http://127.0.0.1:8080", token = null, timeout = 30000) {
    this.baseUrl = String(baseUrl).replace(/\/+$/, "");
    this.token = token || "";
    this.timeout = timeout;
  }

  async _request(method, path, payload = undefined) {
    const headers = { Accept: "application/json" };
    const init = { method, headers };
    if (payload !== undefined) {
      headers["Content-Type"] = "application/json";
      init.body = JSON.stringify(payload);
    }
    if (this.token) headers.Authorization = "Bearer " + this.token;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);
    let resp;
    try {
      resp = await fetch(this.baseUrl + path, { ...init, signal: controller.signal });
    } catch (err) {
      throw new PasmError(
        `无法连接 pasm-framework 服务：${err.message}`, 0,
        "确认服务已启动、地址与端口正确");
    } finally {
      clearTimeout(timer);
    }

    const text = await resp.text();
    let obj = {};
    try { obj = text ? JSON.parse(text) : {}; } catch { obj = { error: text.slice(0, 200) }; }
    if (!resp.ok) {
      throw new PasmError(obj.error || "请求失败", resp.status, obj.hint || "");
    }
    return obj;
  }

  /** 发一条消息，返回回复文本 */
  async chat(text, { sessionId = "default", userId = null, meta = null } = {}) {
    const payload = { text, session_id: sessionId };
    if (userId !== null) payload.user_id = userId;
    if (meta !== null) payload.meta = meta;
    const r = await this._request("POST", "/api/chat", payload);
    return r.reply || "";
  }

  /**
   * 流式发送，逐条产出事件对象（async generator）。
   *
   * 事件：{type:"delta",text} / {type:"replace",text} / {type:"done"} / {type:"error"}
   *
   * 用法：
   *   let acc = "";
   *   for await (const ev of c.chatStream("你好")) {
   *     if (ev.type === "delta") { acc += ev.text; process.stdout.write(ev.text); }
   *     else if (ev.type === "replace") acc = ev.text;   // 护栏改写过 → 整条替换
   *   }
   */
  async *chatStream(text, { sessionId = "default", userId = null, meta = null } = {}) {
    const payload = { text, session_id: sessionId };
    if (userId !== null) payload.user_id = userId;
    if (meta !== null) payload.meta = meta;
    const headers = { "Content-Type": "application/json", Accept: "text/event-stream" };
    if (this.token) headers.Authorization = "Bearer " + this.token;

    let resp;
    try {
      resp = await fetch(this.baseUrl + "/api/chat/stream", {
        method: "POST", headers, body: JSON.stringify(payload),
      });
    } catch (err) {
      throw new PasmError(
        `无法连接 pasm-framework 服务：${err.message}`, 0,
        "确认服务已启动、地址与端口正确");
    }
    if (!resp.ok) {
      const raw = await resp.text();
      let obj = {};
      try { obj = JSON.parse(raw); } catch { obj = { error: raw.slice(0, 200) }; }
      throw new PasmError(obj.error || "请求失败", resp.status, obj.hint || "");
    }
    const reader = resp.body.getReader();
    const decoder = new TextDecoder();
    let buf = "";
    for (;;) {
      const { value, done } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });
      const parts = buf.split("\n\n");
      buf = parts.pop() || "";
      for (const part of parts) {
        const line = part.trim();
        if (!line.startsWith("data:")) continue;
        const chunk = line.slice(5).trim();
        if (!chunk) continue;
        try { yield JSON.parse(chunk); } catch { /* 跳过坏帧 */ }
      }
    }
  }

  /** 批量写入资料库，返回新增条数（需服务端启用 knowledge_base） */
  async ingest(items) {
    const r = await this._request("POST", "/api/ingest", { items });
    return r.added || 0;
  }

  async resetSession(sessionId) {
    const r = await this._request("POST", "/api/sessions/reset", { session_id: sessionId });
    return !!r.ok;
  }

  kbStats() { return this._request("GET", "/api/kb/stats"); }
  plugins() { return this._request("GET", "/api/plugins"); }
  summary() { return this._request("GET", "/api/summary"); }

  /** 健康检查（免鉴权） */
  health() { return this._request("GET", "/healthz"); }
}

module.exports = { PasmClient, PasmError };
